File: src/types/entities.ts
// IMPORTANT NOTE: If changes are made to this file, you must to sync the changes with the rust types module

export type BaseType = {
  id: string
  createdAtTs: number
  updatedAtTs: number
  note: string
  deleted: boolean
}

export type Environment = BaseType & {
  name: string
}
export type EnvironmentCreate = Omit<
  Environment,
  'id' | 'createdAtTs' | 'updatedAtTs' | 'deleted'
>

export type App = BaseType & {
  url: string // js URL type
  environmentId: string
  name: string
  labels: string[]
}
export type AppCreate = Omit<
  App,
  'id' | 'createdAtTs' | 'updatedAtTs' | 'deleted'
>

export type Credential = BaseType & {
  appId: string
  context: string
  username: string
  password: string | null // password can be null if the app doesn't require a password, e.g. API key, mfa app
  url: string | null
}
export type CredentialCreate = Omit<
  Credential,
  'id' | 'createdAtTs' | 'updatedAtTs' | 'deleted'
>

export type Secret = BaseType & {
  appId: string
  key: string
  value: string
}

export type SecretCreate = Omit<
  Secret,
  'id' | 'createdAtTs' | 'updatedAtTs' | 'deleted'
>

export type Certificate = BaseType & {
  name: string
  fileName: string
  fileExtension: string
  value: string
  environmentId: string
  labels: string[]
}
export type CertificateCreate = Omit<
  Certificate,
  'id' | 'createdAtTs' | 'updatedAtTs' | 'deleted'
>

File: src-tauri/src/domain/entities.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize)]
pub struct Environment {
    pub id: String,
    pub created_at_ts: i64,
    pub updated_at_ts: i64,
    pub name: String,
    pub note: String,
    pub deleted: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct App {
    pub id: String,
    pub created_at_ts: i64,
    pub updated_at_ts: i64,
    pub name: String,
    pub url: String,
    pub environment_id: String,
    pub labels: Vec<String>,
    pub note: String,
    pub deleted: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Credential {
    pub id: String,
    pub created_at_ts: i64,
    pub updated_at_ts: i64,
    pub app_id: String,
    pub context: String,
    pub username: String,
    pub password: Option<String>,
    pub url: Option<String>,
    pub note: String,
    pub deleted: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Secret {
    pub id: String,
    pub created_at_ts: i64,
    pub updated_at_ts: i64,
    pub app_id: String,
    pub key: String,
    pub value: String,
    pub note: String,
    pub deleted: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Certificate {
    pub id: String,
    pub created_at_ts: i64,
    pub updated_at_ts: i64,
    pub name: String,
    pub file_name: String,
    pub file_extension: String,
    pub value: String,
    pub environment_id: String,
    pub labels: Vec<String>,
    pub note: String,
    pub deleted: bool,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Configuration {
    pub database_url: String,
    pub database_token: String,
}
